download-test_readline.rb

```diff
@@ -0,0 +1,95 @@
+require 'fileutils'
+
+module HTTPS_DL
+
+  # number of concurrent connections
+  CONNECTIONS = 4
+
+  HOST = "https://raw.githubusercontent.com"
+  URI_GH = URI HOST
+  BASE = "ruby/ruby/master"
+
+  # 1st - ruby/ruby dir  , 2nd - reline dir    , 3rd - filename
+  FILES = [
+    ["test/readline"     , "test/ext/readline" , "helper.rb"                ],
+    ["test/readline"     , "test/ext/readline" , "test_readline.rb"         ],
+    ["test/readline"     , "test/ext/readline" , "test_readline_history.rb" ],
+    ["tool/lib"          , "tool/lib"          , "colorize.rb"              ],
+    ["tool/lib"          , "tool/lib"          , "envutil.rb"               ],
+    ["tool/lib"          , "tool/lib"          , "find_executable.rb"       ],
+    ["tool/lib"          , "tool/lib"          , "leakchecker.rb"           ],
+    ["tool/lib/minitest" , "tool/lib/minitest" , "unit.rb"                  ],
+    ["tool/lib/test"     , "tool/lib/test"     , "unit.rb"                  ],
+    ["tool/lib/test/unit", "tool/lib/test/unit", "assertions.rb"            ],
+    ["tool/lib/test/unit", "tool/lib/test/unit", "core_assertions.rb"       ],
+    ["tool/lib/test/unit", "tool/lib/test/unit", "parallel.rb"              ],
+    ["tool/lib/test/unit", "tool/lib/test/unit", "testcase.rb" ]
+  ]
+
+  class << self
+
+    def run
+      files = FILES
+
+      dirs = FILES.map { |l| l[1] }.uniq
+      dirs.each { |dir| FileUtils.mkdir_p("./#{dir}") unless Dir.exist? dir }
+
+      @connections = []
+      @mutex = Mutex.new
+
+      CONNECTIONS.times do
+        @connections << Thread.new do
+          Net::HTTP.start(URI_GH.host, URI_GH.port, :use_ssl => true, :verify_mode => OpenSSL::SSL::VERIFY_PEER) do |http|
+            done = false
+            loop do
+              path = nil
+              dir = nil
+              file = nil
+              @mutex.synchronize do
+                if files.empty?
+                  done = true
+                else
+                  path, dir, file = files.shift
+                end
+              end
+              break if done
+
+              uri = URI("#{HOST}/#{BASE}/#{path}/#{file}")
+              req = Net::HTTP::Get.new uri.request_uri
+              http.request req do |res|
+                unless Net::HTTPOK === res
+                  STDOUT.puts "Can't download #{path}/#{file} from #{HOST}/#{BASE}/"
+                  exit 1
+                end
+                File.binwrite "./#{dir}/#{file}", res.body
+              end
+            end
+          end
+        end
+      end
+      @connections.each { |th| th.join }
+    end
+
+    def rename_so
+      ext = RbConfig::CONFIG['DLEXT']
+      archdir = RbConfig::CONFIG['archdir']
+      if File.exist? "#{archdir}/readline.#{ext}"
+        Dir.chdir(archdir) { File.rename "readline.#{ext}", "readline.no" }
+        STDOUT.puts "Renamed readline.#{ext}"
+      else
+        STDOUT.puts "Couldn't find readline.#{ext}"
+      end
+    end
+  end
+end
```

Comment:
Downloading test/readline in this ruby/reline repository is just for testing lib/reline. The ext/readline directory and generated readline.so file is tested in [ruby/ruby](https://github.com/ruby/ruby) or [ruby/readline-ext](https://github.com/ruby/readline-ext) repositories. So I think this logic, `HTTP_DL.rename_so` isn't needed.

Reply:
Forgive me, I can't recall.  I thought that if the readline extension file existed, the tests were also run on readline?  I'm probably mistaken, but in the Actions test logs, many of the 'step logs' for this are empty...

Reply:
I just removed the so rename in a fork branch, and the test results are:
```
92 tests, 685 assertions, 0 failures, 0 errors, 2 skips
```

But here, with the rename, the results are:
```
46 tests, 329 assertions, 0 failures, 0 errors, 2 skips
```

So, I think that without the rename, the readline tests are running here, which I assumed was not correct.  Or, a problem with readline extension shouldn't cause a CI failure here...

Reply:
The test/readline always checks whether ext/readline is built and whether lib/reline exists in [test/readline/helper.rb](https://github.com/ruby/ruby/blob/90d082e374abb2f76075ec143b255e3a53157b41/test/readline/helper.rb). Then, [generate a test suit for ext/readline](https://github.com/ruby/ruby/blob/90d082e374abb2f76075ec143b255e3a53157b41/test/readline/test_readline.rb#L780) if it [exists](https://github.com/ruby/ruby/blob/90d082e374abb2f76075ec143b255e3a53157b41/test/readline/test_readline.rb#L787), and [generate a test suit for lib/reline as a readline compatible mode](https://github.com/ruby/ruby/blob/90d082e374abb2f76075ec143b255e3a53157b41/test/readline/test_readline.rb#L789) if it [exists](https://github.com/ruby/ruby/blob/90d082e374abb2f76075ec143b255e3a53157b41/test/readline/test_readline.rb#L804). Those two test suits have the same test methods because those methods are generated by `include BasetestReadline`. So if a repository doesn't have ext/readline, almost half test methods will not be generated. So the results are correct.

Reply:
I understand that.  The four test classes are:
```
TestReadline
TestReadlineHistory

TestRelineAsReadline
TestRelineAsReadlineHistory
```
Do you want to run `TestReadline` & `TestReadlineHistory` here?

1. Rather than rename the readline extension, another solution would be to use an ENV variable to control what test classes run.

2. If the script to download the files is ruby, it could be added to the rake task.

Reply:
> Do you want to run TestReadline & TestReadlineHistory here?

No. I want to test only the lib/reline by test/readline here. Therefore, only `TestRelineAsReadline` is needed here. `TestReadline` should not be run here.

Reply:
Also, `TestRelineAsReadlineHistory` should be run, and `TestReadlineHistory` isn't needed here.

Reply:
Ok, the PR currently only runs the `TestRelineAsReadline` and `TestRelineAsReadlineHistory`.  Each pair of tests has 46 tests.

But, as renaming the readline extension file is unexpected if one is running the tests locally, I thought adding control via ENV would be helpful, as we can add that to the rake task.

See PR https://github.com/ruby/ruby/pull/2916.. If that's accepted, I'll update this PR.

Reply:
The [test/reline/helper.rb](https://github.com/ruby/ruby/blob/90d082e374abb2f76075ec143b255e3a53157b41/test/readline/helper.rb) sets `ReadlineSo` if built ext/readline binary exists, and sets `Reline` if `lib/readline` exists. So `if defined?(ReadlineSo)` and `if defined?(Reline)` are enough to check the necessity of the test suites generation.

Reply:
Ok. I just removed the code to rename the readline ext file.  Now, see:

https://github.com/ruby/reline/runs/461064951?check_suite_focus=true#step:7:9
and
https://github.com/ruby/reline/runs/461064999?check_suite_focus=true#step:6:13

Ubuntu is running 92 tests, and Windows mingw is running 46.

So, the readline tests are running here.  The reason that mingw is running 46 is because I build it so the default is reline, and `readline.so` is renamed `readline.so_`.  mswin is seg faulting with readline because the package doesn't like the encodings you've got here, but it does pass with the encodings in ruby/ruby.  I should probably disable it also...

Sorry for not making this clear.  Is this clear now?

Reply:
Ahhh...I understood what you want to do. I understood `HTTP_DL.rename_so` is for disable *bundled* readline.so. I'm sorry, it's my mistake. So sorry to swing around.

Could you add `HTTP_DL.rename_so` again? I'll merge this Pull Request after you add it.

Reply:
No problem, we're on the same page now.

Renaming extension files isn't a good thing to do, especially if one is running the tests locally.

So, the only other option is to add a way to allow test selection.  See PR https://github.com/ruby/ruby/pull/2916, which works as follows:
```
ENV["TestReadlineOrReline"] undefined:    both tests run (no changes needed to ruby CI)
ENV["TestReadlineOrReline"] = "Readline": readline tests run
ENV["TestReadlineOrReline"] = "Reline":   reline tests run
```

With that (or something like it), reline can run it's tests, readline can do the same, and ruby/ruby can do both?

Reply:
In ruby/ruby, it's in a too complex situation. [Some RubyCI instances](https://rubyci.org/) have ext/readline, some RubyCI instances skip lib/reline tests, some RubyCI instances have ext/readline with Editline and skip many tests...so I added some check logics in test/reline/helper.rb. Now, all RubyCI instances are green. And all instances of RubyCI can't receive any settings including environment variables. So `require` and `rescue LoadError` check is the best way for RubyCI I think.

Reply:
The logic for the PR is kind of different, but I set it up so that Ruby's CI would pass exactly as before with no changes.  IOW, when ENV["TestReadlineOrReline"] is undefined and hence, nil), the logic is exactly as before...

Reply:
I can rebase the PR if you'd like,  I think it was done before https://github.com/ruby/ruby/commit/57c26231ddaee50608aa1c97084b3f3394204612 "Use `brew upgrade` instead of `brew update`", which fixed the macOS fail.

Reply:
OK, I got the picture. If the environment variable exists, we don't need `HTTP_DL.rename_so`. It makes sense. And I reviewed ruby/ruby#2916 first.

Reply:
Thanks.

And we won't irritate people by renaming the extension file...

I can squash the last commit here.  When the Ruby PR is merged, I can force a windows head/master update, but not the others.

Maybe wait until all the ruby-heads have the commit before merging this.  Lastly, in the near future, we should be able to combine the Actions workflow files, instead of having separate OS specific files.  Thanks.